Mathlib/RingTheory/RootsOfUnity/More.lean

```diff
@@ -0,0 +1,79 @@
+/-
+Copyright (c) 2024 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import Mathlib.FieldTheory.KummerExtension
+import Mathlib.RingTheory.RootsOfUnity.Basic
+
+/-!
+# More results on primitive roots of unity
+
+(We put these in a separate file because of the `KummerExtension` import.)
+
+Assume that `μ` is a primitive `n`th root of unity in an integral domain `R`. Then
+$$ \prod_{k=1}^{n-1} (1 - \mu^k) = n \,; $$
+see `IsPrimitiveRoot.prod_one_sub_pow_eq_order` and its variant
+`IsPrimitiveRoot.prod_pow_sub_one_eq_order` in terms of `∏ (μ^k - 1)`.
+
+We use this to deduce that `n` is divisible by `(μ - 1)^k` in `ℤ[μ] ⊆ R` when `k < n`.
+-/
+
+/-- If `A` is an `R`-algebra and `μ : A`, then `μ^k-1` is divisible by `μ-1` in `R[μ] ⊆ A`. -/
+lemma Algebra.adjoin.sub_one_dvd_pow_sub_one (R : Type*) {A : Type*} [CommRing R] [CommRing A]
+    [Algebra R A] (μ : A) (k : ℕ) :
+    ∃ z ∈ adjoin R {μ}, μ ^ k - 1 = z * (μ - 1) := by
+  refine ⟨(Finset.range k).sum (μ ^ ·), ?_, (geom_sum_mul μ k).symm⟩
+  exact Subalgebra.sum_mem _ fun m _ ↦ Subalgebra.pow_mem _ (self_mem_adjoin_singleton _ μ) _
+
+variable {R : Type*} [CommRing R] [IsDomain R]
+
+namespace IsPrimitiveRoot
+
+open Finset Polynomial BigOperators
+
+/-- If `μ` is a primitive `n`th root of unity in `R`, then `∏(1≤k<n) (1-μ^k) = n`.
+(Stated with `n+1` in place of `n` to avoid the condition `n ≠ 0`.) -/
+lemma prod_one_sub_pow_eq_order {n : ℕ} {μ : R} (hμ : IsPrimitiveRoot μ (n + 1)) :
+    ∏ k ∈ range n, (1 - μ ^ (k + 1)) = n + 1 := by
+  have := X_pow_sub_C_eq_prod hμ n.zero_lt_succ (one_pow (n + 1))
+  rw [C_1, ← mul_geom_sum, prod_range_succ', pow_zero, mul_one, mul_comm, eq_comm] at this
+  replace this := mul_right_cancel₀ (Polynomial.X_sub_C_ne_zero 1) this
+  apply_fun Polynomial.eval 1 at this
+  simpa only [mul_one, map_pow, eval_prod, eval_sub, eval_X, eval_pow, eval_C, eval_geom_sum,
+    one_pow, sum_const, card_range, nsmul_eq_mul, Nat.cast_add, Nat.cast_one] using this
+
+/-- If `μ` is a primitive `n`th root of unity in `R`, then `(-1)^(n-1) * ∏(1≤k<n) (μ^k-1) = n`.
+(Stated with `n+1` in place of `n` to avoid the condition `n ≠ 0`.) -/
+lemma prod_pow_sub_one_eq_order {n : ℕ} {μ : R} (hμ : IsPrimitiveRoot μ (n + 1)) :
+    (-1) ^ n * ∏ k ∈ range n, (μ ^ (k + 1) - 1) = n + 1 := by
+  have : (-1 : R) ^ n = ∏ k ∈ range n, -1 := by rw [prod_const, card_range]
+  simp only [this, ← prod_mul_distrib, neg_one_mul, neg_sub, ← prod_one_sub_pow_eq_order hμ]
+
+/-- If `μ` is a primitive `n`th root of unity in `R` and `k < n`, then `n` is divisible
+by `(μ-1)^k` in `ℤ[μ] ⊆ R`. -/
+lemma order_eq_mul_self_sub_one_pow {k n : ℕ} (hn : k < n) {μ : R} (hμ : IsPrimitiveRoot μ n) :
```

Comment:
Again we can use `∣` (and `dvd` in the name).

Reply:
See above. Using "dvd" in the name may still make sense, even though the statement does not mention `Dvd.dvd` explicitly.